lib/bandit/http1/adapter.ex

```diff
@@ -288,8 +243,29 @@ defmodule Bandit.HTTP1.Adapter do
 
   def read_req_body(%__MODULE__{}, _opts), do: raise(Bandit.BodyAlreadyReadError)
 
-  @dialyzer {:no_improper_lists, do_read_chunk: 5}
-  defp do_read_chunk(socket, buffer, body, read_size, read_timeout) do
+  @dialyzer {:no_improper_lists, do_read_content_length_body: 4}
+  defp do_read_content_length_body(socket, buffer, body_remaining, opts) do
+    max_desired_bytes = Keyword.get(opts, :length, 8_000_000)
+
+    if byte_size(buffer) >= max_desired_bytes || body_remaining == 0 do
+      # We can satisfy the read request entirely from our buffer
+      bytes_to_return = min(max_desired_bytes, byte_size(buffer))
+      <<to_return::binary-size(bytes_to_return), rest::binary>> = buffer
+      {:ok, to_return, rest, body_remaining}
+    else
+      # We need to read off the wire
+      bytes_to_read = min(max_desired_bytes - byte_size(buffer), body_remaining)
+
+      with {:ok, iolist} <- read(socket, bytes_to_read, [], opts) do
+        to_return = IO.iodata_to_binary([buffer | iolist])
+        body_remaining = body_remaining - IO.iodata_length(iolist)
```

Comment:
Just keep in mind that `IO.iodata_to_binary` uses the same function under the hood which `IO.iodata_length` calls. So basically the same job is done here twice at cost of extra reduction which will be deducted from a process's budget running this adapter.

Reply:
Good call. Two size calls to calculate the difference will probably be cheaper.